### PR TITLE
Multi property Verification

### PR DESCRIPTION
**Pull request by rafaelsamenezes**
This adds a naive support for verifying all  the claims of a given bound. Basically it sequentially checks for one assertion at a time. Note that, ESBMC's output gets a mess. So in future we should work on making it less verbose.

### CODE REVIEW

File: src/goto-symex/slice.h [hunk 1]
```diff
@@ -36,6 +36,27 @@ class simple_slice : public slicer
...
+ * Claim slicer: remove every other claim of the formula
+ * with the exception of the one to keep.
+ */
+class claim_slicer : public slicer
+{
+public:
+  explicit claim_slicer(const size_t claim_to_keep)
+    : claim_to_keep(claim_to_keep)
+  {
+    if(claim_to_keep == 0)
```
Comment (lucasccordeiro) [inline src/goto-symex/slice.h]: Shall we replace it with `!claim_to_keep`?

File: src/esbmc/bmc.cpp [hunk 2]
```diff
@@ -288,6 +289,30 @@ void bmct::report_trace(
   }
 }
 
+void bmct::report_multi_property_trace(
+  smt_convt::resultt &res,
+  const std::string &msg)
+{
+  assert(options.get_bool_option("base-case"));
```
Comment (lucasccordeiro) [inline src/esbmc/bmc.cpp:306]: Why does it work only for the base case?

File: src/esbmc/bmc.cpp [hunk 3]
```diff
@@ -658,3 +688,58 @@ smt_convt::resultt bmct::run_thread(std::shared_ptr<symex_target_equationt> &eq)
...
+  // As of now, it only makes sense to do this for the base-case
+  assert(
+    options.get_bool_option("base-case") &&
+    "Multi-property only supports base-case");
+  smt_convt::resultt final_result;
+
+  /* TODO: For coverage, We should store and remove all the claims that
+   *       were already verified
+   */
+  for(size_t i = 1; i <= remaining_claims; i++)
```
Comment (lucasccordeiro) [inline src/esbmc/bmc.cpp]: don't we have a way to check all assertions at once with the SMT solvers?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]
It would be nice to have a follow-up PR with a better UI.

File: src/esbmc/bmc.cpp (hunk 3 above)
Comment (rafaelsamenezes) [reply src/esbmc/bmc.cpp]: You mean using solvers API or running all of them in parallel? Either way, it should be possible. I would suggest this going in another PR though.

Comment (lucasccordeiro) [reply src/esbmc/bmc.cpp]: Yes, I believe you should be able to do it via API.

I think this current solution is not good.

Comment (lucasccordeiro) [reply src/esbmc/bmc.cpp]: @fbrausse and @mikhailramalho: what do you think?

File: src/esbmc/bmc.cpp (hunk 2 above)
Comment (rafaelsamenezes) [reply src/esbmc/bmc.cpp:306]: There is nothing preventing other modes. I just didn't implement the messages for them.

As of now, It didn't make sense for me. Maybe we could somehow reuse the unwinding assertions that were proven, that's beyond this PR though.

File: src/esbmc/bmc.cpp [hunk 4]
```diff
@@ -658,3 +688,58 @@ smt_convt::resultt bmct::run_thread(std::shared_ptr<symex_target_equationt> &eq)
...
+        options.get_bool_option("no-slice") &&
+        !options.get_bool_option("compact-trace"))
+        is_compact_trace = false;
+
+      goto_tracet goto_trace;
+      build_goto_trace(eq, runtime_solver, goto_trace, is_compact_trace);
+      // TODO: Replace this with a test-case for coverage!
+      std::string output_file = options.get_option("cex-output");
+
+      if(output_file != "")
```
Comment (mikhailramalho) [inline src/esbmc/bmc.cpp]: isEmpty()?

File: src/esbmc/bmc.cpp [hunk 5]
```diff
@@ -637,7 +662,12 @@ smt_convt::resultt bmct::run_thread(std::shared_ptr<symex_target_equationt> &eq)
         std::shared_ptr<smt_convt>(create_solver("", ns, options));
     }
 
-    return run_decision_procedure(runtime_solver, eq);
+    if(
+      options.get_bool_option("multi-property") &&
+      options.get_bool_option("base-case"))
+      return multi_property_check(eq, result->remaining_claims);
+    else
```
Comment (mikhailramalho) [inline src/esbmc/bmc.cpp:679]: No need for the else here

File: src/goto-symex/slice.cpp [hunk 6]
```diff
@@ -141,3 +141,37 @@ bool simple_slice::run(symex_target_equationt::SSA_stepst &steps)
...
+  for(symex_target_equationt::SSA_stepst::iterator it = steps.begin();
+      it != steps.end();
+      it++)
+  {
+    // just find the next assertion
+    if(it->is_assert())
+    {
+      if(
+        counter++ ==
+        claim_to_keep) // this is the assertion that we should not skip!
```
Comment (mikhailramalho) [inline src/goto-symex/slice.cpp:158]: move the comment up or down to improve code readability?

**Review by mikhailramalho — APPROVED** [review]
Oh, that's really cool! Only some nitpick about the code, otherwise, it LGTM.

Some questions/comments:

1. Do the logs overlap or do we just print several logs/cexs? I wonder if we should use the osyncstream to prevent overlapping.

2. I'd love to see some learning regarding unsat claims in a future patch!

3. Do we have some number from verifying unsat problems in SV-COMP?

File: src/esbmc/bmc.cpp (hunk 3 above)
Comment (mikhailramalho) [reply src/esbmc/bmc.cpp]: You can check the code in `symex_target_equationt::convert` to convert all the assertions at once, it should look something like:
```
    smt_conv.assert_ast(
      smt_conv.make_n_ary(&smt_conv, &smt_convt::mk_or, assertions));
```

File: src/esbmc/bmc.cpp (hunk 2 above)
Comment (lucasccordeiro) [reply src/esbmc/bmc.cpp:306]: Could you please add a TODO here?

File: src/esbmc/bmc.cpp (hunk 5 above)
Comment (lucasccordeiro) [reply src/esbmc/bmc.cpp:679]: Well spotted.

File: src/esbmc/bmc.cpp (hunk 3 above)
Comment (lucasccordeiro) [reply src/esbmc/bmc.cpp]: Given that @rafaelsamenezes has now parallelized the VCs check, I think we could consider @mikhailramalho's suggestions for a future PR.

File: src/esbmc/bmc.cpp [hunk 7]
```diff
@@ -658,3 +694,70 @@ smt_convt::resultt bmct::run_thread(std::shared_ptr<symex_target_equationt> &eq)
...
+      {
+        std::ofstream out(fmt::format("{}-{}", output_file, ce_counter++));
+        show_goto_trace(out, ns, goto_trace);
+      }
+
+      final_result = result;
+    }
+  };
+
+  if(options.get_bool_option("parallel-solving"))
```
Comment (lucasccordeiro) [inline src/esbmc/bmc.cpp]: Can you add some comments here?

**Review by lucasccordeiro — APPROVED** [review]
@rafaelsamenezes: could you please add some regression tests? Can you also run this solution over the SV-COMP benchmarks?

**Comment by rafaelsamenezes** [comment]
This PR is no longer valid. See #1022

**Pull request closed without merging**